File: dataStructures&algorithms/algorithm/leetCode/binaryTree&recursion/hasPathSum/hasPathSum.cpp
#include <iostream>

using namespace std;

struct TreeNode
{
          int val;
          TreeNode *left;
          TreeNode *right;
          TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

// Leetcode-112
class Solution
{
public:
          bool hasPathSum(TreeNode *root, int sum)
          {
                    if (root == NULL)
                              return false;

                    if (root->left == NULL && root->right == NULL)
                              return root->val == sum;

                    if (hasPathSum(root->left, sum - root->val))
                              return true;
                    if (hasPathSum(root->right, sum - root->val))
                              return true;

                    return false;
          }
};

int main()
{

          return 0;
}

File: dataStructures&algorithms/algorithm/leetCode/binaryTree&recursion/hasPathSum/hasPathSum.php
<?php
/**
 * @date: 2022/3/14
 * @createTime: 2:37 PM
 */


// Definition for a binary tree node.
class TreeNode
{
    public $val = null;
    public $left = null;
    public $right = null;

    function __construct($val = 0, $left = null, $right = null)
    {
        $this->val = $val;
        $this->left = $left;
        $this->right = $right;
    }
}

class Solution
{
    function hasPathSum($root, $targetSum)
    {
        if (!isset($root))
            return false;

        if (!isset($root->left) && !isset($root->right))
            return $root->val == $targetSum;

        if ($this->hasPathSum($root->left, $targetSum - $root->val))
            return true;

        if ($this->hasPathSum($root->right, $targetSum - $root->val))
            return true;

        return false;
    }
}